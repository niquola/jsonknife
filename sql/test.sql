create extension jsonknife;

SELECT knife_extract('{"a":{"b": {"c": 5.3}}}', '[["a","b","c"]]'); 

SELECT knife_extract('{"a":[{"b": [{"c": 5.3}, {"c": 6.3}]}, {"b": [{"c": 5.3}, {"c": 6.3}]}]}', '[["a","b","c"]]'); 

SELECT knife_extract('{"a": [1, 2, {"b": {"c": 5.3}}, {"b": {"c": 100}}]}', '[["a", 2 ,"b", "c"]]'); 

SELECT knife_extract('{"name": [{"use": "official", "given": ["a", "b"]}, {"use": "common", "given": ["c", "d"]}]}', '[["name",{"use": "official"},"given"]]');
SELECT knife_extract('{"name": [{"use": "official", "given": ["a", "b"]}, {"use": "official", "given": ["c", "d"]}]}', '[["name",{"use": "official"},"given"]]'); 

SELECT knife_extract('{"name": [{"a": {"b": 1}, "c": "ok"}, {"a": {"b": 2}, "c": "fail"}]}', '[["name", {"a":{"b":1}}, "c"]]');

SELECT knife_extract('{"a":{"b": {"c": 5.3, "d": 6.7}}}', '[["a","b","c"],["a","b","d"]]'); 

SELECT knife_extract_text(
  '{"a": {"b": [{"c": "l", "d": "o"}, {"c": 1, "d": ["b", "o", "k"]}]}}',
  '[["a","b","c"],["a","b","d"]]'
  );


SELECT
'^' ||
array_to_string(
  knife_extract_text(
    '{"a": {"b": [{"c": "l", "d": "o"}, {"c": 1, "d": ["b", "o", "k"]}]}}',
    '[["a","b","c"],["a","b","d"]]'
  ),
  '$^'
)
|| '$'
;

SELECT knife_extract_numeric(
'{"a": 1, "b": [2,3], "c": {"d": 5}}',
'[["a"],["b"],["c", "d"]]'
);

SELECT knife_extract_max_numeric(
'{"a": 1, "b": 2, "c": {"d": 5}}',
'[["a"],["b"],["c", "d"]]'
);

SELECT knife_extract_min_numeric(
'{"a": 1, "b": 2, "c": {"d": 5}}',
'[["a"],["b"],["c", "d"]]'
);

SELECT knife_extract_min_timestamptz('{"a":{"b": {"c": "1980"}}}', '[["a", "b", "c"]]');
SELECT knife_extract_min_timestamptz('{"a":{"b": {"c": "1980-02"}}}', '[["a","b","c"]]');
SELECT knife_extract_min_timestamptz('{"a":{"b": {"c": "1980-02-05"}}}', '[["a","b","c"]]');
SELECT knife_extract_min_timestamptz('{"a":{"b": {"c": "1980-02-05T08"}}}', '[["a","b","c"]]');
SELECT knife_extract_min_timestamptz('{"a":{"b": {"c": "1980-02-05T08:30"}}}', '[["a","b","c"]]');

SELECT knife_extract_max_timestamptz('{"a":{"b": {"c": "1980"}}}', '[["a", "b", "c"]]');
SELECT knife_extract_max_timestamptz('{"a":{"b": {"c": "1980-02"}}}', '[["a","b","c"]]');
SELECT knife_extract_max_timestamptz('{"a":{"b": {"c": "1980-02-05"}}}', '[["a","b","c"]]');
SELECT knife_extract_max_timestamptz('{"a":{"b": {"c": "1980-02-05T08"}}}', '[["a","b","c"]]');
SELECT knife_extract_max_timestamptz('{"a":{"b": {"c": "1980-02-05T08:30"}}}', '[["a","b","c"]]');

SELECT knife_extract_timestamptz('{"a":{"b": {"c": ["1980", "1980-02", "1980-02-05", "1980-02-05T08"]}}}', '[["a", "b", "c"]]');

SELECT knife_date_bound('2005-08-09T13:30:42Z',  'min');
SELECT knife_date_bound('2005-08-09T13:30:42Z',  'max');
SELECT knife_date_bound('2005-08-09T13:30:42+03', 'min');
SELECT knife_date_bound('2005-08-09T13:30:42+03', 'max');

SELECT knife_extract_text('{"resourceType": "Some", "name": [{"use": "official", "given": ["nicola"], "family": ["Ryzhikov"]}, {"use": "common", "given": ["c", "d"]}]}', '[["name","given"],["name","family"]]');

create table patient (id serial, resource jsonb);

insert into  patient (resource)
values
('{"resourceType": "Some", "name": [{"use": "official", "given": ["nicola"], "family": ["Ryzhikov"]}, {"use": "common", "given": ["c", "d"]}]}'),
('{"name": [{"given": ["Nikolai"]}], "resourceType": "Patient"}');

SELECT knife_extract_text(resource, '[["name","given"],["name","family"]]') from patient;
