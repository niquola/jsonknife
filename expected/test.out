create extension jsonknife;
SELECT knife_extract('{"a":{"b": {"c": 5.3}}}', '[["a","b","c"]]'); 
 knife_extract 
---------------
 {5.3}
(1 row)

SELECT knife_extract('{"a":[{"b": [{"c": 5.3}, {"c": 6.3}]}, {"b": [{"c": 5.3}, {"c": 6.3}]}]}', '[["a","b","c"]]'); 
   knife_extract   
-------------------
 {5.3,6.3,5.3,6.3}
(1 row)

SELECT knife_extract('{"a": [1, 2, {"b": {"c": 5.3}}, {"b": {"c": 100}}]}', '[["a", 2 ,"b", "c"]]'); 
 knife_extract 
---------------
 {5.3}
(1 row)

SELECT knife_extract('{"name": [{"use": "official", "given": ["a", "b"]}, {"use": "common", "given": ["c", "d"]}]}', '[["name",{"use": "official"},"given"]]');
   knife_extract   
-------------------
 {"\"a\"","\"b\""}
(1 row)

SELECT knife_extract('{"name": [{"use": "official", "given": ["a", "b"]}, {"use": "official", "given": ["c", "d"]}]}', '[["name",{"use": "official"},"given"]]'); 
           knife_extract           
-----------------------------------
 {"\"a\"","\"b\"","\"c\"","\"d\""}
(1 row)

SELECT knife_extract('{"name": [{"a": {"b": 1}, "c": "ok"}, {"a": {"b": 2}, "c": "fail"}]}', '[["name", {"a":{"b":1}}, "c"]]');
 knife_extract 
---------------
 {"\"ok\""}
(1 row)

SELECT knife_extract('{"a":{"b": {"c": 5.3, "d": 6.7}}}', '[["a","b","c"],["a","b","d"]]'); 
 knife_extract 
---------------
 {5.3,6.7}
(1 row)

SELECT knife_extract_text(
  '{"a": {"b": [{"c": "l", "d": "o"}, {"c": 1, "d": ["b", "o", "k"]}]}}',
  '[["a","b","c"],["a","b","d"]]'
  );
 knife_extract_text 
--------------------
 {l,o,b,o,k}
(1 row)

SELECT
'^' ||
array_to_string(
  knife_extract_text(
    '{"a": {"b": [{"c": "l", "d": "o"}, {"c": 1, "d": ["b", "o", "k"]}]}}',
    '[["a","b","c"],["a","b","d"]]'
  ),
  '$^'
)
|| '$'
;
    ?column?     
-----------------
 ^l$^o$^b$^o$^k$
(1 row)

SELECT knife_extract_numeric(
'{"a": 1, "b": [2,3], "c": {"d": 5}}',
'[["a"],["b"],["c", "d"]]'
);
 knife_extract_numeric 
-----------------------
 {1,2,3,5}
(1 row)

SELECT knife_extract_max_numeric(
'{"a": 1, "b": 2, "c": {"d": 5}}',
'[["a"],["b"],["c", "d"]]'
);
 knife_extract_max_numeric 
---------------------------
                         5
(1 row)

SELECT knife_extract_min_numeric(
'{"a": 1, "b": 2, "c": {"d": 5}}',
'[["a"],["b"],["c", "d"]]'
);
 knife_extract_min_numeric 
---------------------------
                         1
(1 row)

SELECT knife_extract_min_timestamptz('{"a":{"b": {"c": "1980"}}}', '[["a", "b", "c"]]');
 knife_extract_min_timestamptz 
-------------------------------
 Tue Jan 01 00:00:00 1980 PST
(1 row)

SELECT knife_extract_min_timestamptz('{"a":{"b": {"c": "1980-02"}}}', '[["a","b","c"]]');
 knife_extract_min_timestamptz 
-------------------------------
 Fri Feb 01 00:00:00 1980 PST
(1 row)

SELECT knife_extract_min_timestamptz('{"a":{"b": {"c": "1980-02-05"}}}', '[["a","b","c"]]');
 knife_extract_min_timestamptz 
-------------------------------
 Tue Feb 05 00:00:00 1980 PST
(1 row)

SELECT knife_extract_min_timestamptz('{"a":{"b": {"c": "1980-02-05T08"}}}', '[["a","b","c"]]');
 knife_extract_min_timestamptz 
-------------------------------
 Tue Feb 05 08:00:00 1980 PST
(1 row)

SELECT knife_extract_min_timestamptz('{"a":{"b": {"c": "1980-02-05T08:30"}}}', '[["a","b","c"]]');
 knife_extract_min_timestamptz 
-------------------------------
 Tue Feb 05 08:30:00 1980 PST
(1 row)

SELECT knife_extract_max_timestamptz('{"a":{"b": {"c": "1980"}}}', '[["a", "b", "c"]]');
    knife_extract_max_timestamptz    
-------------------------------------
 Wed Dec 31 23:59:59.999999 1980 PST
(1 row)

SELECT knife_extract_max_timestamptz('{"a":{"b": {"c": "1980-02"}}}', '[["a","b","c"]]');
    knife_extract_max_timestamptz    
-------------------------------------
 Fri Feb 29 23:59:59.999999 1980 PST
(1 row)

SELECT knife_extract_max_timestamptz('{"a":{"b": {"c": "1980-02-05"}}}', '[["a","b","c"]]');
    knife_extract_max_timestamptz    
-------------------------------------
 Tue Feb 05 23:59:59.999999 1980 PST
(1 row)

SELECT knife_extract_max_timestamptz('{"a":{"b": {"c": "1980-02-05T08"}}}', '[["a","b","c"]]');
    knife_extract_max_timestamptz    
-------------------------------------
 Tue Feb 05 08:59:59.999999 1980 PST
(1 row)

SELECT knife_extract_max_timestamptz('{"a":{"b": {"c": "1980-02-05T08:30"}}}', '[["a","b","c"]]');
    knife_extract_max_timestamptz    
-------------------------------------
 Tue Feb 05 08:30:59.999999 1980 PST
(1 row)

SELECT knife_extract_timestamptz('{"a":{"b": {"c": ["1980", "1980-02", "1980-02-05", "1980-02-05T08"]}}}', '[["a", "b", "c"]]');
                                                   knife_extract_timestamptz                                                   
-------------------------------------------------------------------------------------------------------------------------------
 {"Tue Jan 01 00:00:00 1980 PST","Fri Feb 01 00:00:00 1980 PST","Tue Feb 05 00:00:00 1980 PST","Tue Feb 05 08:00:00 1980 PST"}
(1 row)

SELECT knife_date_bound('2005-08-09T13:30:42Z',  'min');
       knife_date_bound       
------------------------------
 Tue Aug 09 06:30:42 2005 PDT
(1 row)

SELECT knife_date_bound('2005-08-09T13:30:42Z',  'max');
          knife_date_bound           
-------------------------------------
 Tue Aug 09 06:30:42.999999 2005 PDT
(1 row)

SELECT knife_date_bound('2005-08-09T13:30:42+03', 'min');
       knife_date_bound       
------------------------------
 Tue Aug 09 03:30:42 2005 PDT
(1 row)

SELECT knife_date_bound('2005-08-09T13:30:42+03', 'max');
          knife_date_bound           
-------------------------------------
 Tue Aug 09 03:30:42.999999 2005 PDT
(1 row)

SELECT knife_extract_text('{"resourceType": "Some", "name": [{"use": "official", "given": ["nicola"], "family": ["Ryzhikov"]}, {"use": "common", "given": ["c", "d"]}]}', '[["name","given"],["name","family"]]');
  knife_extract_text   
-----------------------
 {nicola,c,d,Ryzhikov}
(1 row)

create table patient (id serial, resource jsonb);
insert into  patient (resource)
values
('{"resourceType": "Some", "name": [{"use": "official", "given": ["nicola"], "family": ["Ryzhikov"]}, {"use": "common", "given": ["c", "d"]}]}'),
('{"name": [{"given": ["Nikolai"]}], "resourceType": "Patient"}');
SELECT knife_extract_text(resource, '[["name","given"],["name","family"]]') from patient;
  knife_extract_text   
-----------------------
 {nicola,c,d,Ryzhikov}
 {Nikolai}
(2 rows)

